### PR TITLE
planner, cte: fix cte in the embeded union

### LINKED ISSUE #37928: CTE + Union return incorrect results
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. query
```
with cte1 as (select 1), cte2 as (select 2) select * from cte1 union (with cte2 as (select 3) select * from cte2 union all select * from cte2);
```

### 2. What did you expect to see? (Required)

```
+------+
| 1    |
+------+
|    3 |
|    1 |
+------+
2 rows in set (0.00 sec)
```

### 3. What did you see instead (Required)

```
+------+
| 1    |
+------+
|    2 |
|    1 |
+------+
2 rows in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

v6.2.0

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: YangKeao <yangkeao@chunibyo.icu>

### What problem does this PR solve?

Issue Number: close #37928 

Problem Summary:

The `With` in the second element of union `WITH xxx SELECT UNION (WITH xxx SELECT UNION SELECT)` is ignored. It can cause multiple problems like #37928, or the table is not found (as the cte disappeared)...

### What is changed and how it works?

Pass the `With` field in the `buildIntersect` function.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
Fix the issue that the WITH clause in the second elements in the union is ignored.
```

### CODE REVIEW

**Review by CbcWestwolf — APPROVED** [review]

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -7533,3 +7533,24 @@ func TestExplainAnalyzeDMLCommit(t *testing.T) {
 	require.NoError(t, err)
 	tk.MustQuery("select * from t").Check(testkit.Rows())
 }
+
+func TestOverrideCTEInUnion(t *testing.T) {
```
Comment (wjhuang2016) [inline planner/core/integration_test.go:7537]: You can push the test to cmd/explain-test/t/cte

Comment (YangKeao) [reply planner/core/integration_test.go]: Good! I moved this test to cmd/explaintest/t/cte :beers: Thanks!

Comment (YangKeao) [reply planner/core/integration_test.go]: PTAL

**Review by wjhuang2016 — APPROVED** [review]

**Comment by YangKeao** [comment]
@wjhuang2016 @CbcWestwolf Could you help me to merge this PR?

**Comment by CbcWestwolf** [comment]
> Could you help me to merge this PR?

I am not a committer yet. :-|

**Pull request merged**